src/widgets/dropdown_type.h

```diff
@@ -12,61 +12,209 @@
-	void SetDimension(Dimension d);
+/**
+ * Drop down icon component.
+ * @tparam TBase Base component.
+ * @tparam TEnd Position icon at end if true, or start if false.
+ */
+template<class TBase, bool TEnd = false>
+class DropDownIcon : public TBase {
+	SpriteID sprite; ///< Sprite ID to be drawn.
+	PaletteID palette; ///< Palette ID to use.
+	Dimension dim; ///< Bounding box dimensions of sprite.
+	uint sprite_y; ///< Vertical height of sprite, in case
+public:
+	template <typename... Args>
+	explicit DropDownIcon(SpriteID sprite, PaletteID palette, Args&&... args) : TBase(std::forward<Args>(args)...), sprite(sprite), palette(palette)
+	{
+		this->dim = GetSpriteSize(this->sprite);
+		this->sprite_y = dim.height;
+	}
+
+	uint Height() const override { return std::max(this->dim.height, this->TBase::Height()); }
+	uint Width() const override { return this->dim.width + WidgetDimensions::scaled.hsep_normal + this->TBase::Width(); }
+
+	void Draw(const Rect &full, const Rect &r, bool sel, Colours bg_colour) const override
+	{
+		bool rtl = TEnd ^ (_current_text_dir == TD_RTL);
+		DrawSprite(this->sprite, this->palette, r.WithWidth(this->dim.width, rtl).left, CenterBounds(r.top, r.bottom, this->sprite_y));
+		this->TBase::Draw(full, r.Indent(this->dim.width + WidgetDimensions::scaled.hsep_normal, rtl), sel, bg_colour);
+	}
+
+	/**
+	 * Override bounding box dimensions of sprite, to allow multiple options to have consistent spacing.
+	 * @param dim New bounding box to assign.
+	 */
+	void SetDimension(const Dimension &dim)
+	{
+		this->dim = dim;
+	}
 };
 
+/**
+ * Drop down checkmark component.
+ * @tparam TBase Base component.
+ * @tparam TFs Font size.
+ * @tparam TEnd Position checkmark at end if true, or start if false.
+ */
+template<class TBase, bool TEnd = false, FontSize TFs = FS_NORMAL>
+class DropDownCheck : public TBase {
+	bool checked; ///< Is item checked.
+	Dimension dim; ///< Dimension of checkmark.
+public:
+	template <typename... Args>
+	explicit DropDownCheck(bool checked, Args&&... args) : TBase(std::forward<Args>(args)...), checked(checked)
+	{
+		this->dim = GetStringBoundingBox(STR_JUST_CHECKMARK, TFs);
+	}
+
+	uint Height() const override { return std::max<uint>(this->dim.height, this->TBase::Height()); }
+	uint Width() const override { return this->dim.width + WidgetDimensions::scaled.hsep_wide + this->TBase::Width(); }
+
+	void Draw(const Rect &full, const Rect &r, bool sel, Colours bg_colour) const override
+	{
+		bool rtl = TEnd ^ (_current_text_dir == TD_RTL);
+		if (this->checked) {
+			DrawStringMultiLine(r.WithWidth(this->dim.width, rtl), STR_JUST_CHECKMARK, this->GetColour(sel), SA_CENTER, false, TFs);
+		}
+		this->TBase::Draw(full, r.Indent(this->dim.width + WidgetDimensions::scaled.hsep_wide, rtl), sel, bg_colour);
+	}
+};
+
+/* Commonly used drop down list items. */
+using DropDownListDividerItem = DropDownDivider<DropDownListItem>;
+using DropDownListStringItem = DropDownString<DropDownDivider<DropDownListItem>>;
```

Comment:
I don't fully get why `DropDownDivider` as base of `DropDownString`

Reply:
Because the old base DropDownListItem provided the divider itself. The alternative is to undo #11447, and explicitly create a DropDownListDividerItem. That may not be too much hassle now.